### PR TITLE
Support more Bit-Vector primitives in Alt-Ergo's native language

### PR DESCRIPTION
**Pull request by hra687261**

### CODE REVIEW

**Comment by hra687261** [comment]
> If I understand correctly, this PR adds a new builtin theory for the alt-ergo language, and this new theory coincides with the smtlib's one, is that correct ?

Yes!

> but there should be a matching addition to alt-ergo's repo to document that fact.

Yes, it will be added in this PR: https://github.com/OCamlPro/alt-ergo/pull/554

File: src/loop/typer.ml [hunk 1]
```diff
@@ -803,6 +803,13 @@ let non_linear_expression =
     ~hints:[text_hint]
     ~name:"Non linear expression in linear arithmetic logic" ()
 
+let bitvector_app_expected_nat =
+  Report.Error.mk ~code ~mnemonic:"bitvector-app-expected-nat"
+    ~message:(fun fmt t ->
+        Format.fprintf fmt "Expected a natural number as an argument, \
+                            instead got %a" Dolmen_std.Term.print t)
+    ~name:"Bad bitvector application argument type." ()
```
Comment (Gbury) [inline src/loop/typer.ml:811]: It's not the type of the argument that's wrong.

**Review by Gbury — CHANGES_REQUESTED** [review]

File: src/typecheck/bitv.ml [hunk 2]
```diff
@@ -12,13 +12,28 @@ module Ae = struct
 
     type _ Type.err +=
       | Invalid_bin_char : char -> Dolmen.Std.Term.t Type.err
+      | Expected_Nat : Dolmen.Std.Term.t -> Dolmen.Std.Term.t Type.err
```
[the remark this answers is not in the document]
Comment (hra687261) [reply src/typecheck/bitv.ml]: Done.

File: src/typecheck/bitv.ml [hunk 3]
```diff
@@ -12,13 +12,28 @@ module Ae = struct
...
     let parse_binary env s ast =
       match String.iter Misc.Bitv.check_bin s with
       | () -> T.mk s
       | exception Misc.Bitv.Invalid_char c ->
         Type._error env (Ast ast) (Invalid_bin_char c)
 
+    let int_of_term env ast a =
+      match a.Dolmen_std.Term.term with
+      | Symbol { ns = Value Integer; name = Simple name; } ->
+        let i = int_of_string name in
```
[the remark this answers is not in the document]
Comment (hra687261) [reply src/typecheck/bitv.ml:28]: Done.

File: src/loop/typer.ml [hunk 4]
```diff
@@ -803,6 +803,13 @@ let non_linear_expression =
     ~hints:[text_hint]
     ~name:"Non linear expression in linear arithmetic logic" ()
 
+let bitvector_app_expected_nat =
+  Report.Error.mk ~code ~mnemonic:"bitvector-app-expected-nat"
+    ~message:(fun fmt t ->
+        Format.fprintf fmt "Expected a natural number as an argument, \
+                            instead got %a" Dolmen_std.Term.print t)
```
[the remark this answers is not in the document]
Comment (hra687261) [reply src/loop/typer.ml]: Done.

File: src/loop/typer.ml (hunk 1 above)
Comment (hra687261) [reply src/loop/typer.ml]: Fixed

File: src/loop/typer.ml [hunk 5]
```diff
@@ -836,6 +836,15 @@ let non_linear_expression =
...
     ~name:"Non linear expression in linear arithmetic logic" ()
 
+let bitvector_app_expected_nat =
+  Report.Error.mk ~code ~mnemonic:"bitvector-app-expected-nat"
+    ~message:(fun fmt t ->
+        Format.fprintf fmt "Expected a natural number as an argument, \
+                            but instead got the following untyped term:@ %a"
+          Dolmen_std.Term.print t)
+    ~name:"Some bitvector functions in alt-ergo require their first argument \
+           to be a literal natural number" ()
```
Comment (Gbury) [inline src/loop/typer.ml:846]: That's too long  of a name ! I think you confused the name and the hint I suggested. If you want, you can take a look at the other uses of `hints` in this file, for instance : https://github.com/Gbury/dolmen/blob/master/src/loop/typer.ml#L593

File: src/loop/typer.ml [hunk 6]
```diff
@@ -836,6 +836,15 @@ let non_linear_expression =
     ~hints:[text_hint]
     ~name:"Non linear expression in linear arithmetic logic" ()
 
+let bitvector_app_expected_nat =
```
Comment (Gbury) [inline src/loop/typer.ml:839]:
```suggestion
let bitvector_app_expected_nat_lit =
```

File: src/typecheck/bitv.mli [hunk 7]
```diff
@@ -11,8 +11,11 @@ module Ae : sig
 
     type _ Type.err +=
       | Invalid_bin_char : char -> Dolmen.Std.Term.t Type.err
-      (** Error raised when a character that isn't '0' or '1' is
-          used inside a bitvector string *)
+      (** Error raised when a character that isn't '0' or '1' is used inside a
+          literal bitvector. *)
```
Comment (Gbury) [inline src/typecheck/bitv.mli:15]:
```suggestion
          literal bitvector in binary form. *)
```

**Review by Gbury — CHANGES_REQUESTED** [review]

File: src/loop/typer.ml (hunk 5 above)
Comment (hra687261) [reply src/loop/typer.ml:846]: Wow! I can't believe that I pushed that, sorry! I will fix it!

File: src/loop/typer.ml (hunk 1 above)
Comment (Gbury) [reply src/loop/typer.ml:851]:
```suggestion
    ~name:"Bad bitvector application argument" ()
```

File: src/loop/typer.ml [hunk 8]
```diff
@@ -836,6 +841,15 @@ let non_linear_expression =
     ~hints:[text_hint]
     ~name:"Non linear expression in linear arithmetic logic" ()
 
+let bitvector_app_expected_nat_lit =
+  Report.Error.mk ~code ~mnemonic:"bitvector-app-expected-nat"
+    ~message:(fun fmt t ->
+        Format.fprintf fmt "Expected a natural number as an argument, \
```
Comment (Gbury) [inline src/loop/typer.ml:847]:
```suggestion
        Format.fprintf fmt "Expected a natural number literal as an argument, \
```

**Review by Gbury — APPROVED** [review]

**Pull request merged**